### DISCUSSION TITLE
Why not using `#typecheck::` comments instead of `if typing.TYPE_CHECKING`?

### DISCUSSION THREAD

**Discussion by xareelee** [python/typing#1154] [General]
If you need to know what the exact type of a variable is by type inferencing, you need to use `typing.TYPE_CHECKING`:

```python
c: list[int] = [1, 2, 3]
if typing.TYPE_CHECKING:
        print('this line will not be printed')    # this line will not be executed
        reveal_type(c)   # This line only works for mypy
```

Why not using comment style, just like a hint for linter?

```python
c: list[int] = [1, 2, 3]
# typecheck::reveal_type(c)    # mypy reads any comment starting with `typecheck::` and execute it when analyzing.
```

To turn off mypy function call, just make the comment from `# typecheck::` to `#! typecheck::`

```python
c: list[int] = [1, 2, 3]
#! typecheck::reveal_type(c)    # mypy won't read this line
```

Why do we need an `if typing.TYPE_CHECKING` statement for mypy?

**Comment by erictraut** [comment]
The `reveal_type` call was recently added to the runtime. You can import it from `typing_extensions`, and it will also be available in `typing` as of Python 3.11. Note that `reveal_type` is not specific to mypy; it works with all popular Python type checkers.

If you use a type checker in your code editor rather than running it as a command-line tool, you can get "live" feedback about a symbol's type simply by hovering over it with your mouse pointer. Mypy wasn't designed for use a language server, but pyright (the type checker I work on) is designed for this use. This eliminates the need for using `reveal_type`.

I'll also note that I generally recommend against using `typing.TYPE_CHECKING` in your code because it "lies" to the type checker and potentially obscures runtime errors that would otherwise be caught.

**Reply by xareelee** [reply]
> The `reveal_type` call was recently added to the runtime. You can import it from `typing_extensions`, and it will also be available in `typing` as of Python 3.11. Note that `reveal_type` is not specific to mypy; it works with all popular Python type checkers.

May I ask what `reveal_type`  will do at runtime? Do I need to remove them in production code?

I remember mypy and many other linters/checkers support comment syntax annotation. I think this feature is not hard to implement.

Using comment annotation ensures the code will not be execute at runtime, and you don't need to use `import` statement to use that. This is coding friendly for other developers.

**Comment by JelleZijlstra** [comment]
`if TYPE_CHECKING` has other usages than `reveal_type()`. For `reveal_type()` specifically, you can instead use `from typing_extensions import reveal_type` (or `from typing import reveal_type` on 3.11) to avoid the condition.

I wasn't around when `if TYPE_CHECKING` was designed, but it has a couple of advantages over a magic comment: it uses normal Python syntax, so it's easier to understand for other tools, and it naturally supports more than one line of code.

**Reply by Akuli** [reply]
`if TYPE_CHECKING` is also super easy to implement. Type checkers have to understand things like `if foo is not None` or `if sys.platform == "win32"` anyway, and `if TYPE_CHECKING` is just like those, but even simpler.

**Comment by xareelee** [comment]
Currently, I write `reveal_type` using one line code as the following:

```python
(reveal_type(c) if typing.TYPE_CHECKING else None)
```

Does this statement have any issue in production?

**Reply by erictraut** [reply]
`reveal_type` simply returns the value it was passed. This allows it to be used within other expressions and without the use of `typing.TYPE_CHECKING`.
